File: src/main/java/ru/arc/sync/base/ActionCanceller.java
package ru.arc.sync.base;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Future;
import java.util.function.Consumer;

import org.bukkit.event.Event;

import static ru.arc.util.Logging.debug;

public class ActionCanceller {

    final Map<Class<? extends Event>, Consumer<Event>> cancellers = new ConcurrentHashMap<>();
    Map<UUID, Long> preventUntil = new ConcurrentHashMap<>();
    Future<?> pruneTask;

    public ActionCanceller() {
    }

    public <T extends Event> void registerCanceller(Class<T> clazz, Consumer<T> canceller) {
        cancellers.put(clazz, (Consumer<Event>) canceller);
    }

    public boolean checkAndCancel(UUID playerUuid, Event event) {
        Long until = preventUntil.get(playerUuid);
        if (until != null) {
            if (until > System.currentTimeMillis()) {
                debug("Prevented event " + event.getClass().getSimpleName() + " for " + playerUuid);
                debug("Prevented until " + until);
                debug("Current time " + System.currentTimeMillis());
                var canceller = cancellers.get(event.getClass());
                if (canceller != null){
                    canceller.accept(event);
                    return true;
                }
                return false;
            } else {
                preventUntil.remove(playerUuid);
            }
        }
        return false;
    }

    public void addTemporaryProtection(UUID uuid, long ms) {
        preventUntil.put(uuid, System.currentTimeMillis() + ms);
    }

    public void stopTasks() {
        if (pruneTask != null && !pruneTask.isCancelled()) pruneTask.cancel(true);
    }


}

File: src/main/kotlin/ru/arc/sync/base/ActionCanceller.kt
package ru.arc.sync.base

import org.bukkit.event.Event
import ru.arc.util.Logging.debug
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Future

class ActionCanceller {

    val cancellers: MutableMap<Class<out Event>, (Event) -> Unit> = ConcurrentHashMap()
    val preventUntil: MutableMap<UUID, Long> = ConcurrentHashMap()
    var pruneTask: Future<*>? = null

    fun <T : Event> registerCanceller(clazz: Class<T>, canceller: (T) -> Unit) {
        @Suppress("UNCHECKED_CAST")
        cancellers[clazz] = canceller as (Event) -> Unit
    }

    fun checkAndCancel(playerUuid: UUID, event: Event): Boolean {
        val until = preventUntil[playerUuid]
        if (until != null) {
            if (until > System.currentTimeMillis()) {
                debug("Prevented event {} for {}", event.javaClass.simpleName, playerUuid)
                debug("Prevented until {}", until)
                debug("Current time {}", System.currentTimeMillis())
                val canceller = cancellers[event.javaClass]
                if (canceller != null) {
                    canceller(event)
                    return true
                }
                return false
            } else {
                preventUntil.remove(playerUuid)
            }
        }
        return false
    }

    fun addTemporaryProtection(uuid: UUID, ms: Long) {
        preventUntil[uuid] = System.currentTimeMillis() + ms
    }

    fun stopTasks() {
        pruneTask?.takeIf { !it.isCancelled }?.cancel(true)
    }
}
